File: App/LinoWriting/Models/AgentLog.swift
import Foundation

public struct AgentLog: Codable, Equatable, Identifiable, Sendable, Hashable {
    public let id: String
    public let chapterId: String?
    public let agentName: String
    public let inputPreview: String?
    public let outputPreview: String?
    public let latencyMs: Int?
    public let tokensIn: Int?
    public let tokensOut: Int?
    public let error: String?
    public let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case chapterId = "chapter_id"
        case agentName = "agent_name"
        case inputPreview = "input_preview"
        case outputPreview = "output_preview"
        case latencyMs = "latency_ms"
        case tokensIn = "tokens_in"
        case tokensOut = "tokens_out"
        case error
        case createdAt = "created_at"
    }
}

File: Backend/app/models/agent_log.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base
from app.models.common import utc_now


class AgentLog(Base):
    __tablename__ = "agent_logs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    chapter_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("chapters.id", ondelete="SET NULL"))
    agent_name: Mapped[str] = mapped_column(String(32), nullable=False)
    input_preview: Mapped[str | None] = mapped_column(Text)
    output_preview: Mapped[str | None] = mapped_column(Text)
    latency_ms: Mapped[int | None] = mapped_column(Integer)
    tokens_in: Mapped[int | None] = mapped_column(Integer)
    tokens_out: Mapped[int | None] = mapped_column(Integer)
    error: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, nullable=False)
